### PR TITLE
pbft: allow replicas to complain about primary

### PR DESCRIPTION
**Pull request by corecode**
This addresses #796.  Applies on top of #1265.

### CODE REVIEW

**Comment by corecode** [comment]
Signed-off-by: Simon Schubert sis@zurich.ibm.com

**Comment by corecode** [comment]
this behave test passes locally

**Comment by jyellick** [comment]
@corecode Could you write a brief overview of how the complaints system functions?

My (admittedly rough, and probably incorrect) understanding, is that a peer keeps track of the requests it receives (and then broadcasts to the network), and will broadcast a complaint to the network when it believes the primary is censoring its request.  When a peer receives a complaint, it re-submits the request to the primary, and then if the primary censors the request, sends a view change.  Is this correct / how is it wrong?

So a few immediate questions.  Why do I see complaints at the plugin specific level instead of in a common way for all plugins? I see this being done in Sieve and batch, but not classic, I assume this is intentional, but why?

I've admittedly not been the best about this, but @kchristidis pointed out to me that some of my checkins were missing function documentation and were generally not meeting go lint/vet standards.  Could you make sure they pass for your new files? (We'll need to do a wholesale linting of the other files, but do not want to create merge headaches).  Some documentation might also help make the review easier.

**Comment by corecode** [comment]
Yes, your description is correct.  The reason why it is not in the pbft core is that single requests get forwarded to the primary, and the pbft core only sees a batch and not separate requests.  I did not implement the functionality for classic pbft yet, because batch with size set to 1 will provide the same performance and functionality, so immediate support for complaints did not seem necessary.  More broadly, I think we could remove classic in favor of batch with size=1.

I'll add function documentation in a followup commit.

**Comment by srderson** [comment]
@corecode Could you rebase this PR? There is a merge conflict. Thanks

**Comment by corecode** [comment]
@srderson Rebased, ready to merge.

File: consensus/obcpbft/custodian/custodian.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+type custody struct {
+	id       string
+	data     interface{}
+	deadline time.Time
+	canceled bool
+}
+
+type CustodyPair struct {
```
Comment (srderson) [inline consensus/obcpbft/custodian/custodian.go]: Could you run golint on this file please. You will see warnings for missing comments on public types and functions. Thanks

**Comment by srderson** [comment]
Thanks. LGTM

**Comment by corecode** [comment]
As indicated in #1445, this PR contains a race condition in one of the unit tests (`TestBatchCustody`).

I believe the cause for the failed test is that as complaints get submitted and trigger a view change, the request order is changed, which leads to the older request being discarded as stale.  However, replica 0 (previous leader) still has a PBFT request filed, but due to the test configuration, cannot communicate this (can't send messages).  Therefore PBFT stays in view change.

**Pull request merged**